### ISSUE TITLE
2.0 changes

### ISSUE THREAD

**Issue by mpilquist** [scodec/scodec-bits#141]
What changes are folks interested in for a potential scodec-bits 2.0? The last binary breakage was appx 4 years ago. Stuff I'm considering:

- going back to `Int` based indexing & sizes instead of today's `Long` based indexing. We switched to long based indexing in order to support mmap-ing large files, but these days, there are much better options for working with files (e.g. fs2-io).
- supporting zero-copy scenarios for converting to arrays?
- adding a dependency on cats-kernel? Or perhaps instead integrating with the 2.13 collection library? Sort of an arbitrary thing but I'd like to get monoid instances without imports / compat libraries.
- Dotty support - specifically, porting the hex/bin string interpolators

What else?

Not sure about timeframe for a 2.0 -- probably along with cats-effect 3.0 and fs2 3.0.

**Comment by rossabaker** [comment]
In favor of cats-kernel:
* Has been more stable recently than cats-collections
* Can avoid the `cats.implicits._` import (minor: many files need it anyway)
* No compromises to fit any non-FP features of scala-collections

In favor of scala-collections integration:
* Just changed, so maybe it's settled and Cats is overdue
* Might be able to span breaking changes with scala_x.y folders
* Even careful dependencies like Cats are liabilities

**Comment by hamnis** [comment]
+1 for scala collections and dotty support and no extra dependencies

**Comment by barambani** [comment]
I'd vote for dotty support as well

**Comment by pm47** [comment]
+1 for going back to Int based indexing & supporting zero-copy scenarios for converting to arrays

**Comment by soronpo** [comment]
Hi, sorry for coming in late, but here are my current uses for scodec-bits that I had to wrap around it due to lack of direct support:
* Bit Ordering. Indeed it shouldn't really matter how the internals are in terms of bit ordering, but when the user accesses a specific bit by index, when the user inputs/outputs a string equivalent encoding of the value, or when the user concatenates vectors, it matters a great deal.
For example: 
I use Msbit is at the left down-to Lsbit at the right. 
Current behavior:
```scala
bin"11" ++ bin"0000" 
val res5: scodec.bits.BitVector = BitVector(6 bits, 0xc0)
```
I expect this to be  `BitVector(6 bits, 0x20)`.
```scala
bin"10000".set(0)
val res6: scodec.bits.BitVector = BitVector(5 bits, 0x80)
```
I expect this to be `BitVector(5 bits, 0x11)`
So essentially I propose adding a Bit-Ordering implicit argument to all related functions with a default implicit available in scope to provide the current behavior for backward comparability.
* Compile-time length type argument: 
Dotty provides type operations (and in Scala 2 via singleton-ops).
So if we have also a (limited) cluster of functions that can support `Sized.BitVector[L]`
We can have `BitVector[2] + BitVector[3]` to produce a `BitVector[5]`.

**Comment by mpilquist** [comment]
@soronpo For bit ordering, I'm not a fan of implicit configuration - especially when there's a default parameter in play. My gut reaction is to try building an explicit view that supports the reverse ordering. E.g something like `bin"10000".inverseIndexing.set(0).defaultIndexing`. Or maybe such a view could support both bit and byte ordering - `hex"deadbeef".bits.view(ByteOrdering.BigEndian, BitOrdering.LSB).slice(4, 16).force`

> I expect this to be BitVector(6 bits, 0x30).

Note this is the result of calling `toHex` on a vector with size not evenly divisible by 8. The decision to left pad was discussed on Gitter but I'm having trouble finding it right now. Surprisingly, there's no built-in combinator for left padding a bit vector to nearest byte boundary - e.g. `padLeft((size + 7) / 8)`

**Comment by soronpo** [comment]
I don't understand problem of having an implicit parameter that can also be introduced globally for the project. It's very similar to a language flag. For my use-case I need everything default "the-other-way". Seems a real bother to create an opaque type/value class/tagging just to reverse all the behavior.

**Comment by mpilquist** [comment]
Implicit params with defaults are way too error prone in my experience. That's how we originally passed around a thread pool back in early days of `scalaz.Task` (ala `Strategy`). Maintaining code bases that used it was super frustrating as folks would refactor some code and forget to carry along the implicit, resulting in the default instance continually sneaking back in to the code base. It was a never ending game of whack-a-mole.

Over time, I've settled on a code style that eschews implicit params for just about everything except type classes. I've found the inconvenience of added syntax is worth the clarity. YMMV of course. :)

**Comment by soronpo** [comment]
The default concept is to ease migration. You can also introduce a default with a deprecation and after a while remove the default entirely. So it cannot be confusing if the compiler prevents the code from compiling and instructs the user to introduce the proper import via an `implicitNotFound` error message.

**Comment by BardurArantsson** [comment]
Please do not add dependencies other than the standard library. It's a bit part of why my company is using this library in the first place.

(We're in a position where we have to impose scodec-bits on consumers of our code, so an absolutely minimal dependency 'footprint' is a huge plus for us since we really don't want to get into the business of imposing transitive dependencies on them.)

**Comment by mpilquist** [comment]
New plan is to stick with 1.1.x series and not break bincompat at this time. Dotty support has been backported to 1.1.x (and released under 1.1.18).
